Add a new gas cost model (#508)
gavofyork/graypaper: text/pvm.tex

Reviewer: Not sure this is really true any more?

Reviewer 2: > limited to being at most 16

If the maximum legal instruction length is 16, why is 24 used as the upper bound in the equation above? Shouldn't it be 15 instead?

Author: The 24 upper bound is for the *skip* function, i.e. the distance to the next instruction.

The "implicit though limited to being at most 16" essentially means "the whole thing is designed so that you only ever need to read at most 16 bytes per instruction to read all of the instruction's operands", or in more practical terms - you can read a single `u128` per instruction and efficiently decode it through a series of bit operations instead of having to read byte-by-byte.

Note: the distance to the next instruction (i.e. the *skip* function) doesn't necessarily have to be the same as the number of bytes used by the opcode + the arguments. (For example, we take advantage of this property to encode a dispatch table at the beginning of JAM blobs to keep the entry points for is_authorized/accumulate/refine at nice round addresses, where the instructions' *skip* is wider than it is necessary.)

If the "However, each instruction [...]" part is confusing I suppose we could delete it? (As it doesn't really hold any normative power per se.)

Reviewer 2: > Note: the distance to the next instruction (i.e. the skip function) doesn't necessarily have to be the same as the number of bytes used by the opcode + the arguments. 

Alright. I wrongly assumed that was the case; that is, one immediately follows the other. This clarifies the situation.

\newcommand*{\deblob}{\text{deblob}}
\newcommand*{\smod}{\text{smod}}
\newcommand*{\rtz}{\text{rtz}}
\newcommand*{\gascostforblock}{\gascounter^{\Delta}}
\newcommand*{\startofbasicblock}{\mathfrak{L}}

We declare the general \textsc{pvm} function $\Psi$. We assume a single-step invocation function define $\Psi_1$ and define the full \textsc{pvm} recursively as a sequence of such mutations up until the single-step mutation results in a halting condition. We additionally define the function $\deblob$ which extracts the instruction data, opcode bitmask and dynamic jump table from a \textsc{pvm} program blob:
\begin{align}
  \Psi&\colon \abracegroup{

>     \tuple{\blob, \pvmreg, \gas, \bool, \regs, \ram} &\to \tuple{\set{\halt, \panic, \oog} \cup \set{\fault, \host} \times \pvmreg, \pvmreg, \signedgas, \bool, \regs, \ram}\\

Reviewer: Think we don't need `\signedgas` here any more? As the behaviour now is that gas is not consumed by regular execution if it would drop below 0. We still need signed gas higher up for host calls.

    \tup{\pvm¬blob, \imath, \gascounter, \gaschargedflag, \registers, \mem} &\mapsto \begin{cases}
      \Psi(\pvm¬blob, \imath', \gascounter', \gaschargedflag', \registers', \mem') &\when \varepsilon = \continue\\

>       \tup{\oog, \imath, \gascounter', \gaschargedflag', \registers, \mem} &\when \gascounter' < 0\\

Reviewer: Think this case can just be removed now? `\gascounter'` is never <0 AFAICT but OOG will be handled fine by the last case here?

\newcommand*{\deblob}{\text{deblob}}
\newcommand*{\smod}{\text{smod}}
\newcommand*{\rtz}{\text{rtz}}

We declare the general \textsc{pvm} function $\Psi$. We assume a single-step invocation function define $\Psi_1$ and define the full \textsc{pvm} recursively as a sequence of such mutations up until the single-step mutation results in a halting condition. We additionally define the function $\deblob$ which extracts the instruction data, opcode bitmask and dynamic jump table from a program blob:
\newcommand*{\gascostforblock}{\gascounter^{\Delta}}
\newcommand*{\startofbasicblock}{\mathfrak{L}}
We declare the general \textsc{pvm} function $\Psi$. We assume a single-step invocation function define $\Psi_1$ and define the full \textsc{pvm} recursively as a sequence of such mutations up until the single-step mutation results in a halting condition. We additionally define the function $\deblob$ which extracts the instruction data, opcode bitmask and dynamic jump table from a \textsc{pvm} program blob, validates its structure, and verifies whether the given $\imath$ is a valid instruction counter location within the program:
\begin{align}
  \Psi&\colon \abracegroup{

>     \tuple{\blob, \pvmreg, \gas, \regs, \ram} &\to \tuple{\set{\halt, \panic, \oog} \cup \set{\fault, \host} \times \pvmreg, \pvmreg, \signedgas, \regs, \ram}\\

Reviewer: The signature of `\Psi` is also given in the overview section, should replace `\signedgas` with `\gas` there. There is some text talking about negative gas there which is now not really true.

      \Psi_1(\mathbf{c}, \mathbf{k}, \mathbf{j}, \imath, \gascounter, \gaschargedflag, \registers, \mem) &\when \tup{\mathbf{c}, \mathbf{k}, \mathbf{j}} = \deblob(\pvm¬blob, \imath) \\
      \tup{\panic, \imath, \gascounter, \gaschargedflag, \registers, \mem} &\otherwise
    \end{cases}
  }\\
  \deblob&\colon\abracegroup{
    (\blob, \pvmreg) &\to \tuple{\blob, \bitstring, \sequence{\pvmreg}} \cup \error \\
    (\pvm¬blob, \imath) &\mapsto \begin{cases}
      \tup{\mathbf{c}, \mathbf{k}, \mathbf{j}} &\when \exists!\,\mathbf{c}, \mathbf{k}, \mathbf{j} : \pvm¬blob = \encode{\len{\mathbf{j}}} \concat \encode[1]{z} \concat \encode{\len{\mathbf{c}}} \concat \encode[z]{\mathbf{j}} \concat \encode{\mathbf{c}} \concat \encode{\mathbf{k}}\,,\ \mathfrak{v}_{\text{blob}}(\mathbf{c}, \mathbf{k}, 0) \land \mathfrak{v}_{\text{loc}}(\mathbf{c}, \mathbf{k}, \imath) \\
      \error &\otherwise
    \end{cases} \\
    \where &\mathfrak{v}_{\text{blob}}\colon\abracegroup{
      \tuple{\blob, \bitstring, \pvmreg} &\to \bool \\
      (\mathbf{c}, \mathbf{k}, \imath) &\mapsto \begin{cases}
        \top &\when \imath > 0 \land \imath = \len{\mathbf{k}}\\

>         \mathfrak{v}(\mathbf{c}, \mathbf{k}, \imath + 1 + \text{skip}(\imath)) &\otherwhen \mathfrak{v}_{\text{loc}}(\mathbf{c}, \mathbf{k}, \imath)\\

Reviewer: ```suggestion
        \mathfrak{v}_{\text{blob}}(\mathbf{c}, \mathbf{k}, \imath + 1 + \text{skip}(\imath)) &\otherwhen \mathfrak{v}_{\text{loc}}(\mathbf{c}, \mathbf{k}, \imath)\\
```

    (\pvm¬blob, \imath) &\mapsto \begin{cases}
      \tup{\mathbf{c}, \mathbf{k}, \mathbf{j}} &\when \exists!\,\mathbf{c}, \mathbf{k}, \mathbf{j} : \pvm¬blob = \encode{\len{\mathbf{j}}} \concat \encode[1]{z} \concat \encode{\len{\mathbf{c}}} \concat \encode[z]{\mathbf{j}} \concat \encode{\mathbf{c}} \concat \encode{\mathbf{k}}\,,\ \mathfrak{v}_{\text{blob}}(\mathbf{c}, \mathbf{k}, 0) \land \mathfrak{v}_{\text{inst}}(\mathbf{c}, \mathbf{k}, \imath) \\
      \error &\otherwise
    \end{cases} \\
  }{
    \\[0.2pt]
    \where \mathfrak{v}_{\text{blob}}\colon&\abracegroup{
      \tuple{\blob, \bitstring, \pvmreg} &\to \bool \\
      (\mathbf{c}, \mathbf{k}, \imath) &\mapsto \begin{cases}
        \top &\when \imath > 0 \land \imath = \len{\mathbf{k}}\\
        \bot &\otherwhen \imath + 1 + \text{skip}(\imath) = \len{\mathbf{k}} \land \mathbf{c}_\imath \not\in T\\
        \mathfrak{v}_{\text{blob}}(\mathbf{c}, \mathbf{k}, \imath + 1 + \text{skip}(\imath)) &\otherwhen \mathfrak{v}_{\text{inst}}(\mathbf{c}, \mathbf{k}, \imath)\\
        \bot &\otherwise
      \end{cases}
    }\\
    \mathfrak{v}_{\text{inst}}\colon&\abracegroup{
      \tuple{\blob, \bitstring, \pvmreg} &\to \bool \\

>       (\mathbf{c}, \mathbf{k}, \imath) &\mapsto \len{\mathbf{k}} = \len{\mathbf{c}} \land \imath < \len{\mathbf{k}} \land \mathbf{k}_{\imath} = 1 \land \mathbf{c}_\imath \in U

Reviewer 2: Minor nit: this does not affect the overall semantics, but perhaps placing |**k**| = |**c**| at the top level, to be evaluated once, would be more appropriate.

Author: This was deliberate, as on the original line (i.e. the one with `\exists!\,\mathbf{c} ...`) there is no space to fit this, and I didn't want to overcomplicate/uglify the equations, so the easiest thing to do was to just move it to `v_inst` (since, as you've noted, it doesn't matter semantically whether it's there or at toplevel).

But I'd be happy to move it if you can suggest a way to do it nicely. (:

Reviewer 2: As mentioned, this does not affect the semantics. If moving it makes the result less readable, keep it as is. This is fine

  \basicblocks \equiv \left(\set{0} \cup \set{\build{n + 1 + \Fskip(n)}{n \in \Nmax{\len{\mathbf{c}}} \wedge \mathbf{k}\sub{n} = 1 \wedge \mathbf{c}\sub{n} \in T}}\right) \cap \set{\build{n}{\mathbf{k}\sub{n} = 1 \wedge \mathbf{c}\sub{n} \in U}}
\end{equation}

For convenience we will define function $\startofbasicblock$ which represents the start of a basic block for any $\imath$ within that basic block. Formally:

\begin{equation}

>   \startofbasicblock\colon \abracegroup{

Reviewer: I think there are a few cases here where things aren't well-defined. eg If we start executing at 0 but there isn't a valid instruction at 0 then the gas charging stuff doesn't seem well-defined -- there is no basic block at 0, so what is charged? I wonder if more generally we should require that the initial program counter is at the start of a basic block; I wouldn't be surprised if some implementations required this now, but it doesn't seem to be a requirement here.

Author: > I wonder if more generally we should require that the initial program counter is at the start of a basic block

I'd love to be able to do that, unfortunately we can't require this as we also need the ability to resume in the middle of a basic block (e.g. for an inner VM which previously was interrupted due to a page fault).

That said, jumps *within* the program are guaranteed to be valid and can only go to the beginning of basic blocks, so it's mostly a question of what to do on the initial entry. Hmmm...

I think you're right in that some of the teams are most likely handling these corner cases differently.

I suppose maybe the most prudent thing to do right now would be to just prevalidate the program before we start execution, and reject it if it contains any abnormalities? That would be relatively simple to do and unambiguous. For JAM 1.0 at very least we'll be assuming that everyone's going to be running a whole program recompiler anyway, which implies that we'll be paying O(n) cost anyway per unique program invocation (no fancy recompiler/interpreter hybrids nor lazy JIT recompilation) which means that such prevalidation would effectively be "free". And we could always relax this in the future.

Thoughts?

Reviewer: > That said, jumps within the program are guaranteed to be valid and can only go to the beginning of basic blocks, so it's mostly a question of what to do on the initial entry. Hmmm...

Yes, I meant initial entry. Resumption seems easier as you only care about the "native" PC for that? At least I don't think there's any way to set or inspect the PC of a VM?

Prevalidation sounds reasonable to me. As you say it shouldn't really cost anything with a recompiler and would close off some attack surface.

Author: > At least I don't think there's any way to set or inspect the PC of a VM?

Indeed. You can only set the program counter when you start the VM (with `invoke`), and you can only fetch it when you end it (with `expunge`). Otherwise it's inaccessible to keep things simple.

\begin{equation}
  \tup{\tilde{\varepsilon}, \gascounter', \gaschargedflag'} = \begin{cases}
    \tup{\continue, \gascounter, \top} &\when \gaschargedflag = \top \\
    \tup{\continue, \gascounter - \gascostforblock(\mathbf{c}, \mathbf{k}, \startofbasicblock(\imath)), \top} &\otherwhen \gascounter \ge \gascostforblock(\mathbf{c}, \mathbf{k}, \startofbasicblock(\imath)) \\
    \tup{\oog, \gascounter, \bot} &\otherwise
  \end{cases}
\end{equation}

We also have to adjust the state based on the posterior value of the instruction counter, to force a gas charge on the next step when necessary:

\begin{equation}
  \gaschargedflag^* = \begin{cases}

>     \bot &\when \gaschargedflag' = \bot \lor \imath' < \imath \lor \startofbasicblock(\imath') \ne \startofbasicblock(\imath) \\

Reviewer: I think you want <= rather than < for the restarted-block condition? As presumably a jump/branch can jump to itself. I guess we want to avoid triggering this in the case of a fault, where we don't advance over the instruction. I _think_ this is fine as we're using `\imath'` rather than `\imath^*` here. It's all a bit subtle though and maybe there is a better way to formalise this...

Author: Hmm... maybe it'd be better to delete this and have the control flow instructions explicitly reset the gas charged flag?

Reviewer: Sounds reasonable. Or maybe rather than detecting the PC changing, check if (a) the opcode is in the "ends basic block" set and (b) was actually executed?

    \top &\otherwise
  \end{cases}
\end{equation}

During the course of executing instructions \textsc{ram} may be accessed. When an index of \textsc{ram} below $2^{16}$ is required, the machine always panics immediately without further changes to its state regardless of the apparent (in)accessibility of the value. Otherwise, should the given index of \textsc{ram} not be accessible then machine state remains unchanged and the exit reason is a fault with the lowest inaccessible \emph{page address} to be read. Similarly, where \textsc{ram} must be mutated and yet mutable access is not possible, then machine state is unchanged, and the exit reason is a fault with the lowest page address to be written which is inaccessible.

> Formally, let $\mathbf{r}$ and $\mathbf{w}$ be the set of indices by which $\mem$ must be subscripted for inspection and mutation respectively in order to calculate the result of $\Psi_1$. We define the memory-access exceptional execution state $\varepsilon^\mu$ which shall, if not $\continue$, singly effect the returned return of $\Psi_1$ as following:

Reviewer: This paragraph could do with rewording a bit I think. `\varepsilon^\mu` doesn't actually exist AFAICT and this equation now also covers out-of-gas.

Perhaps it would be neatest to actually define `\varepsilon^\mu` (as either `\continue`, `\panic`, or `\fault`) and then separately define the `*` variables based on `\varepsilon^\mu` and `\tilde{\varepsilon}`?

\begin{align}
  \using \mathbf{x} &= \set{\build{x}{x \in \mathbf{r} \wedge x \bmod 2^{32} \not\in \readable\mem\ \vee\ x \in \mathbf{w} \wedge x \bmod 2^{32} \not\in \writable\mem}} \\
  \tup{\varepsilon^*, \imath^*, \gascounter^*, \registers^*, \mem^*} &= \begin{cases}
    \tup{\oog, \imath, \gascounter', \registers, \mem} &\when \tilde{\varepsilon} = \oog \\

>     \tup{\varepsilon, \imath', \gascounter', \registers', \mem'} &\when \mathbf{x} = \emset \\

Reviewer: Think you ought to change these `\when` to `\otherwhen` now?

  \basicblocks \equiv \left(\set{0} \cup \set{\build{n + 1 + \Fskip(n)}{n \in \Nmax{\len{\mathbf{c}}} \wedge \mathbf{k}\sub{n} = 1 \wedge \mathbf{c}\sub{n} \in T}}\right) \cap \set{\build{n}{\mathbf{k}\sub{n} = 1 \wedge \mathbf{c}\sub{n} \in U}}
\end{equation}

For convenience we will define $\startofbasicblock(\imath)$ which represents the start of a basic block for any $\imath$ within that basic block. Formally:

\begin{equation}
  \startofbasicblock(\imath) = \max(j \in \basicblocks \mid j \le \imath)
\end{equation}

\subsection{Single-Step State Transition}

We must now define the single-step \textsc{pvm} state-transition function $\Psi_1$:
\begin{equation}
  \Psi_1\colon \abracegroup{
    \tuple{\blob, \bitstring, \sequence{\pvmreg}, \pvmreg, \gas, \bool, \regs, \ram} &\to \tuple{\set{\panic, \halt, \continue } \cup \set{\fault, \host} \times \pvmreg, \pvmreg, \signedgas, \bool, \regs, \ram}\\
    \tup{\pvm¬blob, \mathbf{k}, \mathbf{j}, \imath, \gascounter, \gaschargedflag, \registers, \mem} &\mapsto \tup{\varepsilon^*, \imath^*, \gascounter^*, \gaschargedflag^*, \registers^*, \mem^*}
  }
\end{equation}

> On the very first step of execution, and every time the execution enters a new basic block or jumps back to the beginning of the current basic block, the gas counter of the machine is updated according to the gas cost $\gascostforblock{\startofbasicblock(\imath)}$~(\ref{eq:gascostforblock}) of the target basic block. No instruction is allowed to execute within a basic block without first paying the gas cost of the whole block. In case there's not enough gas remaining to cover the full gas cost, the execution is interrupted and the gas counter remains unchanged. Formally:

Reviewer 2: ```suggestion
On the very first step of execution, and every time the execution enters a new basic block or jumps back to the beginning of the current basic block, the gas counter of the machine is updated according to the gas cost $\gascostforblock{\startofbasicblock(\imath)}$~(\ref{eq:gascostforblock}) of the target basic block. No instruction is allowed to execute within a basic block unless the gas cost for the entire basic block has been charged in advance. In case there's not enough gas remaining to cover the full gas cost, the execution is interrupted and the gas counter remains unchanged. Formally:
```

Just a personal preference to express the same concept, feel free to ignore

\begin{equation}
  \begin{aligned}
    \varepsilon^{\mu} &= \begin{cases}
      \continue &\when \mathbf{x} = \emptyset \\
      \panic &\when \min(\mathbf{x}) \bmod 2^{32} < 2^{16} \\
      \fault \times \Cpvmpagesize\floor{\min(\mathbf{x}) \bmod 2^{32} \div \Cpvmpagesize} &\otherwise
    \end{cases}\\
    \where \mathbf{x} &= \set{\build{x}{x \in \mathbf{r} \wedge x \bmod 2^{32} \not\in \readable\mem\ \vee\ x \in \mathbf{w} \wedge x \bmod 2^{32} \not\in \writable\mem}}
  \end{aligned}
\end{equation}

We define the final execution state, the value of the instruction counter, the values of the registers and the memory as follows:
\begin{align}
  \tup{\varepsilon^*, \imath^*, \registers^*, \mem^*} &= \begin{cases}
    \tup{\varepsilon^{\gascounter}, \imath, \registers, \mem} &\when \varepsilon^{\gascounter} \neq \continue \\
    \tup{\varepsilon^{\mu}, \imath, \registers, \mem} &\otherwhen \varepsilon^{\mu} \neq \continue \\
    \tup{\varepsilon, \imath', \registers', \mem'} &\otherwise
  \end{cases}
\end{align}

> We also have to adjust the state based on the posterior value of the instruction counter, to force a gas charge on the next step when necessary:

Reviewer: ```suggestion
We also have to adjust the state based on the executed instruction, to force a gas charge on the next step when necessary:
```

(Posterior value of instruction counter is no longer inspected)

\begin{equation}
  \gaschargedflag^* = \begin{cases}

>     \bot &\when \gaschargedflag' = \bot \lor (\mathbf{c}_\imath \in T \land \varepsilon^* \in \set{\continue} \cup \set{\host} \times \pvmreg) \\

Reviewer: ```suggestion
    \bot &\when \gaschargedflag' = \bot \lor (\instructions_\imath \in T \land \varepsilon^* \in \set{\continue} \cup \set{\host} \times \pvmreg) \\
```

Think you need `\instructions` here rather than `\mathbf{c}` as we can execute past the end? Possibly as part of validation you could require that the final instruction is a basic-block terminator, that would make this case go away.

  102&\token{count\_set\_bits\_32}&$\displaystyle\reg'_D = \sum_{i = 0}^{31}\bitsfunc{4}(\reg_A \bmod 2^{32})\sub{i}$\\ \mrule
  103&\token{leading\_zero\_bits\_64}&$\displaystyle\reg'_D = \max(n \in \Nmax{65})\ \where \sum_{i = 0}^{i < n} \revbitsfunc{8}(\reg_A)\sub{i} = 0$\\ \mrule
  104&\token{leading\_zero\_bits\_32}&$\displaystyle\reg'_D = \max(n \in \Nmax{33})\ \where \sum_{i = 0}^{i < n} \revbitsfunc{4}(\reg_A \bmod 2^{32})\sub{i} = 0$\\ \mrule
  105&\token{trailing\_zero\_bits\_64}&$\displaystyle\reg'_D = \max(n \in \Nmax{65})\ \where \sum_{i = 0}^{i < n} \bitsfunc{8}(\reg_A)\sub{i} = 0$\\ \mrule
  106&\token{trailing\_zero\_bits\_32}&$\displaystyle\reg'_D = \max(n \in \Nmax{33})\ \where \sum_{i = 0}^{i < n} \bitsfunc{4}(\reg_A \bmod 2^{32})\sub{i} = 0$\\ \mrule
  107&\token{sign\_extend\_8}&$\reg'_D = \unsigned{\signedn{1}{\reg_A \bmod 2^8}}$\\ \mrule
  108&\token{sign\_extend\_16}&$\reg'_D = \unsigned{\signedn{2}{\reg_A \bmod 2^{16}}}$\\ \mrule
  109&\token{zero\_extend\_16}&$\reg'_D = \reg_A \bmod 2^{16}$\\ \mrule
  110&\token{reverse\_bytes}&$\forall i \in \N_8 : \encode[8]{\reg'_D}\sub{i} = \encode[8]{\reg_A}_{7-i}$\\

> %  10X&\token{}&$\reg'_D = ...$\\ \mrule

Reviewer 2: ```suggestion
```

Dead code?

An extended version of the \textsc{pvm} invocation which is able to progress an inner \emph{host-call} state-machine in the case of a host-call halt condition is defined as $\Psi_H$:
\begin{align}
  &\Psi_H^*\colon \abracegroup{
    \tuple{\begin{aligned}
      &\blob, \pvmreg, \gas, \bool, \regs,\\&\ram, \contextmutator{X}, X
    \end{aligned}
    }
    &\to
    \tuple{\set{\panic, \oog, \halt} \cup \set{\fault} \times \pvmreg, \pvmreg, \signedgas, \bool, \regs, \ram, X}\\
    \tup{\pvm¬blob, \imath, \gascounter, \gaschargedflag, \registers, \mem, f, \mathbf{x}} &\mapsto \begin{cases}

>       \multicolumn{2}{l}{\text{let }(\varepsilon', \imath', \gascounter', \gaschargedflag', \registers', \mem') = \Psi(\pvm¬blob, \imath, \gascounter, \gaschargedflag, \registers, \mem):} \\[8pt]

Reviewer 3: Yes, it isn't recursion into `ΨH`. It's about calling the basic PVM with the flag: `Ψ` takes 5 params (without gas-charged flag) and `Ψ*` takes 6 (with the flag). Shouldn't we call `Ψ*` here since we want to carry the flag for per-block metering? (And also adjust (A.1) to use `Ψ*` for "continue" case)

An extended version of the \textsc{pvm} invocation which is able to progress an inner \emph{host-call} state-machine in the case of a host-call halt condition is defined as $\Psi_H$:
\begin{align}
  &\Psi_H^*\colon \abracegroup{
    \tuple{\begin{aligned}

>       &\blob, \pvmreg, \gas, \regs,\\&\ram, \bool, \contextmutator{X}, X

Reviewer 2: ```suggestion
      &\blob, \pvmreg, \gas, \bool, \regs,\\&\ram, \contextmutator{X}, X
```

An extended version of the \textsc{pvm} invocation which is able to progress an inner \emph{host-call} state-machine in the case of a host-call halt condition is defined as $\Psi_H$:
\begin{align}
  &\Psi_H^*\colon \abracegroup{
    \tuple{\begin{aligned}

>       &\blob, \pvmreg, \gas, \bool, \regs,\\&\ram, \contextmutator{X}, X

Reviewer 2: ```suggestion
      &\blob, \pvmreg, \gas, \bool,\\&\regs, \ram, \contextmutator{X}, X
```

Nit. To shift all of this slightly toward the LHS.

We thus define the standard JAM program blob format $\jam¬blob$, which includes not only the raw \textsc{pvm} program blob $\pvm¬blob$, but also information on the state of the \textsc{ram} at program start. Given JAM program blob $\jam¬blob$ and argument data $\mathbf{a}$, we can decode the \textsc{pvm} program blob $\pvm¬blob$, registers $\registers$, and \textsc{ram} $\mem$ by invoking the standard initialization function $Y(\jam¬blob, \mathbf{a})$:
\begin{equation}
Y\colon\abracegroup{
  \tuple{\blob, \blob[:\Cpvminitinputsize]} &\to \tuple{\blob, \regs, \ram}? \\
  \tup{\jam¬blob, \mathbf{a}} &\mapsto \begin{cases}

>     \tup{\pvm¬blob, \registers, \mem} &\when \exists! \tup{\pvm¬blob, \mathbf{o}, \mathbf{w}, z, s} \text{ which satisfy equation \ref{eq:conditions}}\\

Reviewer 2: When this is rendered, it references equation A.42 only 

<img width="680" height="90" alt="Image" src="https://github.com/user-attachments/assets/340fdf9f-2c73-4807-a106-110c1f4c1df8" />

<img width="762" height="67" alt="Image" src="https://github.com/user-attachments/assets/fb58bae0-6778-4196-8632-05e14794f15c" />

But perhaps A.45 should be referenced as well?

<img width="651" height="34" alt="Image" src="https://github.com/user-attachments/assets/91b8bb54-1bd5-4759-b0df-5742d2aa9856" />

This is strange as looks like the label does indeed encompass both. Perhaps this requires some more LaTeX magic

\end{equation}

Note that the first tuple item is the amount of gas consumed by the operation, but never greater than the amount of gas provided for the operation.

\subsection{Gas Cost Model}\label{sec:gascostmodel}

\newcommand*{\rob¬state}{\vec{s}}
\newcommand*{\rob¬cyclesleft}{\vec{c}}
\newcommand*{\rob¬deps}{\vec{p}}
\newcommand*{\rob¬regs}{\vec{r}}
\newcommand*{\rob¬execunits}{\vec{x}}
\newcommand*{\ctr¬cycles}{\dot{c}}
\newcommand*{\ctr¬instructions}{\dot{n}}
\newcommand*{\remaining¬decodeslots}{\dot{d}}
\newcommand*{\remaining¬startpercycle}{\dot{e}}
\newcommand*{\remaining¬execunits}{\dot{x}}
\newcommand*{\cost¬decodeslots}{\check{d}}
\newcommand*{\cost¬cycles}{\check{c}}
\newcommand*{\cost¬execunits}{\check{x}}
\newcommand*{\inst¬dstregs}{\check{r}}
\newcommand*{\inst¬srcregs}{\check{s}}
\newcommand*{\state¬decoding}{1}
\newcommand*{\state¬waiting}{2}
\newcommand*{\state¬executing}{3}
\newcommand*{\state¬executed}{4}
\newcommand*{\ALU}{\text{A}}
\newcommand*{\LOAD}{\text{L}}
\newcommand*{\STORE}{\text{S}}
\newcommand*{\MUL}{\text{M}}
\newcommand*{\DIV}{\text{D}}
\newcommand*{\find¬readytostart}{\mathfrak{S}}

> \newcommand*{\emptyorderedmap}{\langle\,\rangle}

Reviewer: If you're going to define a new notation you should add it to the notation section. I expect no new notation is needed here though, you can probably use either sequences or dictionaries for the reorder buffer?

\newcommand*{\const¬maxdecodepercycle}{4}
\newcommand*{\const¬maxstartpercycle}{5}

The gas cost model for the \textsc{pvm} is a simplified model of a modern CPU microarchitecture, heavily inspired by what's used by production-grade compilers to predict how much time a given piece of code will take. For each basic block in the program the model simulates its execution flow and computes the required number of virtual CPU cycles that would be needed to execute it.

The gas cost $\gascostforblock{\imath}$ for a given basic block starting at instruction opcode index $\imath \in \basicblocks$ is defined by the number of virtual CPU cycles as determined by the the gas cost model transition function, up until every instruction of the basic block it has ingested has been retired and the simulation has converged.

The initial state $\Xi_0$ of the simulation is defined as follows:

\begin{equation}
  \Xi_0(\imath) = \abracegroup{

>     \imath^{(0)} &= \imath \\

Reviewer: The notation here is a bit confusing. What's the purpose of the superscripts on the state variables? They seem to just duplicate the overall state `\Xi` subscript. Is it so the individual variables at a particular step can be referenced? Why not just write eg `\Xi_n(i)_c` instead of `c^(n)`? This would be more consistent with the notation used elsewhere and defined in the notation section.

In any case it would be helpful here to define the type of the state `\Xi_n` and explain the meaning of each of the fields, as is done for most of the other tuples in the GP.

Author: > Yes. I don't think we should invent new notation just for this section, we should stick to what is already defined in section 3.

Okay, fair enough. In my defense, I was operating under the assumption that I should maximize the readability (or at least what I *think* maximizes readability, which, to be fair, is subjective so not everyone might agree) and that a very slight deviation for this section (which is otherwise somewhat complex and self-contained) would be okay.

I'll rewrite it to match the existing conventions better.

    \ctr¬cycles^{(0)} &= 0,
    \ctr¬instructions^{(0)} = 0,
    \remaining¬decodeslots^{(0)} = \const¬maxdecodepercycle,
    \remaining¬startpercycle^{(0)} = \const¬maxstartpercycle \\

>     \rob¬state^{(0)} &= \emptyorderedmap,

Reviewer: It might be clearer to make the reorder buffer a sequence-of-tuples rather than a tuple-of-sequences? Seems like ndot is not really necessary; you could just use the length of this sequence?

    \rob¬cyclesleft^{(0)} = \emptyorderedmap,
    \rob¬deps^{(0)} = \emptyorderedmap,
    \rob¬regs^{(0)} = \emptyorderedmap,

>     \rob¬execunits^{(0)} = \emptyorderedmap \\

Reviewer: It would be helpful here to define the type of the state `\Xi_n` and explain the meaning of each of the fields, as is done for most of the other tuples in the GP.

>     \remaining¬execunits^{(0)} &= \{

Reviewer: This should use tuple notation: `\tup{\is{\ALU}{4}, \is{\LOAD}{4}, ...}`. Same applies to the overall definition of `\Xi`.

      \ALU = 4,
      \LOAD = 4,
      \STORE = 4,
      \MUL = 1,
      \DIV = 1
    \}
  }
\end{equation}

The state transition function of the simulation from step $n$ to step $n + 1$ is defined as follows:

\begin{equation}

>   \Xi_{n + 1}(\Xi_n, \mathbf{c}, \mathbf{k}) = \begin{cases}

Reviewer: If `\mathbf{c}` and `\mathbf{k}` are clearly defined earlier and used as free variables throughout the PVM section then it's probably fine to just use them here. In that case though I would avoid binding them in any functions you define. If you do that you're essentially shadowing them? And I would maybe still add a bit of text to clarify that they continue to have the same meanings as earlier.

>     \Xi_{n + 1}'(\Xi_n, \mathbf{c}, \mathbf{k}) &\when n = 0 \lor (\imath^{(n)} \neq \none \land \cost¬decodeslots(\mathbf{c}, \mathbf{k}, \imath^{(n)}) \leq \remaining¬decodeslots^{(n)} \land |\rob¬state^{(n)}| < 32) \\

Reviewer: The `n = 0 \lor` seems unnecessary here?

    \rob¬state^{(n + 1)}_{\ctr¬instructions^{(n)}} &= \state¬decoding \\
    \rob¬cyclesleft^{(n + 1)}_{\ctr¬instructions^{(n)}} &= \cost¬cycles(\mathbf{c}, \mathbf{k}, \imath^{(n)}) \\
    \rob¬execunits^{(n + 1)}_{\ctr¬instructions^{(n)}} &= \cost¬execunits(\mathbf{c}, \mathbf{k}, \imath^{(n)}) \\
    \rob¬deps^{(n + 1)}_{\ctr¬instructions^{(n)}} &= \set{\build{j}{ \inst¬srcregs(\mathbf{c}, \mathbf{k}, \imath^{(n)}) \cap \rob¬regs^{(n)}_{j} \neq \emptyset }} \\
    j \in \mathbb{N} \Rightarrow \rob¬regs^{(n + 1)}_j &= \begin{cases}
      \inst¬dstregs(\mathbf{c}, \mathbf{k}, \imath^{(n)}) &\when j = \ctr¬instructions^{(n)} \\
      \rob¬regs^{(n)}_j \setminus \inst¬dstregs(\mathbf{c}, \mathbf{k}, \imath^{(n)}) &\otherwise
    \end{cases} \\
  }
\end{equation}

The state transition function $\Xi_{n + 1}''$ which starts the execution of the next pending instruction is defined as follows:

\begin{equation}
  \Xi_{n + 1}''(\Xi_n) = \abracegroup{
    \rob¬state^{(n + 1)}_{\find¬readytostart(\Xi_n)} &= \state¬executing \\
    \remaining¬execunits^{(n + 1)} &= \remaining¬execunits^{(n)} - \rob¬execunits^{(n)}_{\find¬readytostart(\Xi_n)} \\

>     \remaining¬startpercycle^{(n + 1)} &= \remaining¬startpercycle^{(n)} - 1

Author: Yes, this is correct. The idea is to handle that through decode costs and execution unit requirements.

This counter is mostly to cap the number of times (per cycle) we have to check the reorder buffer for the next instruction to start executing. Currently I've set `const¬maxstartpercycle` to 5, but this is somewhat arbitrary - there could actually be more than 5 instructions waiting to start in the ROB (assuming their resource requirements don't overlap), however such situation is rare enough in practice that it doesn't make sense to waste the compute (i.e. to slow down the gas cost model simulation and regress our compile times) on checking this (very heavy diminishing returns; I measured, and picked the 5 based on those measurements).

    \ctr¬instructions^{(0)} = 0,
    \remaining¬decodeslots^{(0)} = \const¬maxdecodepercycle,
    \remaining¬startpercycle^{(0)} = \const¬maxstartpercycle \\
    \rob¬state^{(0)} &= \emptyorderedmap,
    \rob¬cyclesleft^{(0)} = \emptyorderedmap,
    \rob¬deps^{(0)} = \emptyorderedmap,
    \rob¬regs^{(0)} = \emptyorderedmap,
    \rob¬execunits^{(0)} = \emptyorderedmap \\
    \remaining¬execunits^{(0)} &= \{
      \ALU = 4,
      \LOAD = 4,
      \STORE = 4,
      \MUL = 1,
      \DIV = 1
    \}
  }
\end{equation}

The state transition function of the simulation from step $n$ to step $n + 1$ is defined as follows:

\begin{equation}
  \Xi_{n + 1}(\Xi_n, \mathbf{c}, \mathbf{k}) = \begin{cases}
    \Xi_{n + 1}'(\Xi_n, \mathbf{c}, \mathbf{k}) &\when n = 0 \lor (\imath^{(n)} \neq \none \land \cost¬decodeslots(\mathbf{c}, \mathbf{k}, \imath^{(n)}) \leq \remaining¬decodeslots^{(n)} \land |\rob¬state^{(n)}| < 32) \\
    \Xi_{n + 1}''(\Xi_n) &\when \find¬readytostart(\Xi_n) \neq \none \land \remaining¬startpercycle^{(n)} > 0 \\
    \none &\when \imath^{(n)} = \none \land |\rob¬state^{(n)}| = 0 \\
    \Xi_{n + 1}'''(\Xi_n) &\otherwise \\
  \end{cases}
\end{equation}

> Assuming $m \text{ such that } \Xi_{m + 1} = \none$ the final gas cost of a basic block starting at $\imath$ is defined as:

Reviewer 2: This is not clear to me. How is the output of $\Xi_{m + 1} = ∅$?

    \ctr¬instructions^{(n + 1)} &= \ctr¬instructions^{(n)} + 1 \\
    \rob¬state^{(n + 1)}_{\ctr¬instructions^{(n)}} &= \state¬decoding \\
    \rob¬cyclesleft^{(n + 1)}_{\ctr¬instructions^{(n)}} &= \cost¬cycles(\mathbf{c}, \mathbf{k}, \imath^{(n)}) \\
    \rob¬execunits^{(n + 1)}_{\ctr¬instructions^{(n)}} &= \cost¬execunits(\mathbf{c}, \mathbf{k}, \imath^{(n)}) \\
    \rob¬deps^{(n + 1)}_{\ctr¬instructions^{(n)}} &= \set{\build{j}{ \inst¬srcregs(\mathbf{c}, \mathbf{k}, \imath^{(n)}) \cap \rob¬regs^{(n)}_{j} \neq \emptyset }} \\
    j \in \mathbb{N} \Rightarrow \rob¬regs^{(n + 1)}_j &= \begin{cases}
      \inst¬dstregs(\mathbf{c}, \mathbf{k}, \imath^{(n)}) &\when j = \ctr¬instructions^{(n)} \\
      \rob¬regs^{(n)}_j \setminus \inst¬dstregs(\mathbf{c}, \mathbf{k}, \imath^{(n)}) &\otherwise
    \end{cases} \\
  }
\end{equation}

The state transition function $\Xi_{n + 1}''$ which starts the execution of the next pending instruction is defined as follows:

\begin{equation}
  \Xi_{n + 1}''(\Xi_n) = \abracegroup{
    \rob¬state^{(n + 1)}_{\find¬readytostart(\Xi_n)} &= \state¬executing \\

>     \remaining¬execunits^{(n + 1)} &= \remaining¬execunits^{(n)} - \rob¬execunits^{(n)}_{\find¬readytostart(\Xi_n)} \\

Author: Yes, but it's not saturated per-se (because this will not trigger if there are not enough execution units available).

\newcommand*{\const¬maxdecodepercycle}{4}
\newcommand*{\const¬maxstartpercycle}{5}

% Helpers
\newcommand*{\rob¬entry}[2]{#1\subb{#2}}
\newcommand*{\rob¬entry¬field}[3]{\rob¬entry{#1}{#2}_{#3}}
\newcommand*{\in¬rob}{{\gas¬sim¬state_{\rob}}}
\newcommand*{\in¬rob¬entry}[1]{{\in¬rob\subb{#1}}}
\newcommand*{\in¬rob¬field}[2]{\in¬rob¬entry{#1}_{#2}}

% This is like an \abracegroup, but with two independently aligned chunks.
% It makes it possible to have a nicer looking abracegroup blocks where a single
% overly long column doesn't mess up the spacing of the rest of it.
\newcommand{\abracegrouptwo}[2]{%
  \left\{%
  \begin{array}{@{}l@{}}
    \begin{aligned}
      #1
    \end{aligned}
    \\[14pt]
    \hspace{8pt}\begin{aligned}
      #2
    \end{aligned}
  \end{array}
  \right.
}

The gas cost model for the \textsc{pvm} is a simplified model of a modern CPU microarchitecture, heavily inspired by what's used by production-grade compilers to predict how much time a given piece of code will take. For each basic block in the program the model simulates its execution flow and computes the required number of virtual CPU cycles that would be needed to execute it.

> The gas simulation state type $\gas¬sim¬ty$ contains the current instruction counter $\imath$, a cycle counter $\ctr¬cycles$, the number of decode slots $\remaining¬decodeslots$ remaining in the current cycle, the maximum number of instructions $\remaining¬startpercycle$ which are still allowed to start execution in the current cycle, the remaining available execution units $\remaining¬execunits$, and the reorder buffer $\rob$. Formally:

Reviewer: ```suggestion
A gas simulation state, of set $\gas¬sim¬ty$, contains the current instruction counter $\imath$, a cycle counter $\ctr¬cycles$, the number of decode slots $\remaining¬decodeslots$ remaining in the current cycle, the maximum number of instructions $\remaining¬startpercycle$ which are still allowed to start execution in the current cycle, the remaining available execution units $\remaining¬execunits$, and the reorder buffer $\rob$. Formally:
```

And similarly elsewhere; the GP doesn't generally refer to these sets as types.

Author: I was basing this off of this snippet:

```
The inner \textsc{pvm} invocation host-calls, meanwhile, depend on an integrated \textsc{pvm} type, which we shall denote $\pvmguest$. It holds a \textsc{pvm} program blob, instruction counter and \textsc{ram}:
\begin{equation}\label{eq:pvmguest}
  \pvmguest \equiv \tuple{\isa{\pg¬code}{\blob}, \isa{\pg¬ram}{\ram}, \isa{\pg¬pc}{\pvmreg}}
\end{equation}
```

which *does* seem to explicitly call such a type tuple a "type" and enumerates what it contains? Do you want me to change that one too?

\begin{equation}
  \gas¬sim¬ty \equiv \tuple{

>     \isa{\imath}{\set{\pvmreg, \none}},

Reviewer: Think you want `\pvmreg \cup \set{\none}` here.

    \isa{\rob}{\sequence{\rob¬entry¬ty}}
  }\\
\end{equation}

The reorder buffer entry type $\rob¬entry¬ty$ contains its current state $\rob¬state$, the number of cycles left $\rob¬cyclesleft$, a set of reorder buffer indices considered its dependencies $\rob¬deps$, a set of clobbered registers $\rob¬regs$, and the execution units used $\rob¬execunits$. Formally:

\begin{equation}
  \rob¬entry¬ty \equiv \tuple{
    \isa{\rob¬state}{\rob¬state¬ty},
    \isa{\rob¬cyclesleft}{\N},
    \rob¬deps \subseteq \N,
    \rob¬regs \subseteq \Nmax{13},
    \isa{\rob¬execunits}{\execunits¬ty}
  }\\
\end{equation}

The type $\execunits¬ty$ is a tuple which maps each execution unit kind (${\ALU, \LOAD, \STORE, \MUL, \DIV}$) into a number:

\begin{equation}
  \execunits¬ty \equiv \tuple{
    \isa{\ALU}{\N},
    \isa{\LOAD}{\N},
    \isa{\STORE}{\N},
    \isa{\MUL}{\N},
    \isa{\DIV}{\N}
  }
\end{equation}

For convenience we define addition and subtraction of two values of type $\execunits¬ty$ to be equivalent to the memberwise operation of the same kind; formally:

\begin{equation}
\begin{aligned}
  \forall a,b \in \execunits¬ty;\forall \oplus \in \set{+, -}&: a \oplus b \equiv c\\

>   \where \forall k \in \set{\ALU, \LOAD, \STORE, \MUL, \DIV}&:c_k \equiv a_k \oplus b_k\\

Reviewer: I guess there should really be a `c \in \execunits¬ty` somewhere here.

\end{aligned}
\end{equation}

Similarly, a comparison between two values of type $\execunits¬ty$ is true if and only if the chosen comparison holds for all corresponding members:

\begin{equation}
\begin{aligned}
  &\forall a,b \in \execunits¬ty;\forall \oplus \in \set{<, >, \le, \ge}: \\
  &\quad a \oplus b \equiv \bigwedge_{k \in \set{\ALU, \LOAD, \STORE, \MUL, \DIV}} a_k \oplus b_k
\end{aligned}
\end{equation}

> We define function $\inst¬srcregs(\mathbf{c}, \mathbf{k}, \imath)$ which returns a set of source registers read by a given instruction at $\imath$, and the function $\inst¬dstregs(\mathbf{c}, \mathbf{k}, \imath)$ which returns a set of destination registers which are written by a given instruction, as described by the equations in \ref{sec:instructiontables}. This is regardless of whether those registers would actually have been modified by that instruction when executed at runtime. \token{ecalli} is assumed to neither read nor write to any registers in this model.

Reviewer: ```suggestion
We define the function $\inst¬srcregs(\mathbf{c}, \mathbf{k}, \imath)$ which returns the set of source registers read by the instruction at $\imath$, and the function $\inst¬dstregs(\mathbf{c}, \mathbf{k}, \imath)$ which returns the set of destination registers which are written by a given instruction, as described by the equations in \ref{sec:instructiontables}. This is regardless of whether those registers would actually have been modified by that instruction when executed at runtime. \token{ecalli} is assumed to neither read nor write to any registers in this model.
```

> We also define function $\cost¬cycles(\mathbf{c}, \mathbf{k}, \imath)$ which returns the number of cycles a given instruction needs to finish execution, $\cost¬decodeslots(\mathbf{c}, \mathbf{k}, \imath)$ which returns the number of decoding slots necessary to decode it, and $\cost¬execunits(\mathbf{c}, \mathbf{k}, \imath)$ which returns the number of virtual CPU execution units required to start its execution. These simply return the values as specified in \ref{sec:gascosttables}.

Reviewer: We also define the function

The gas cost of a given basic block starting at instruction opcode index $\imath \in \basicblocks$ and given the instruction data $\mathbf{c}$ and the opcode bitmask $\mathbf{k}$ is defined by the number of virtual CPU cycles as determined by the gas cost model transition function, up until every instruction of the basic block it has ingested has been retired and the simulation has converged. Formally:

\begin{equation}\label{eq:gascostforblock}
  \gascostforblock\colon \abracegrouptwo{
    \tup{\blob, \bitstring, \pvmreg} &\to \N\\
    \tup{\mathbf{c}, \mathbf{k}, \imath} &\mapsto \max(\gas¬sim¬state^{\text{final}}_{\ctr¬cycles} - 3, 1)\\
  }{
    \where &\gas¬sim¬state^{\text{final}} &=&\text{ }\gas¬sim(\mathbf{c}, \mathbf{k}, \gas¬sim¬state^{\text{init}})\\

>       &\gas¬sim¬state^{\text{init}} &=&\text{ }\tup{\imath, 0, \const¬maxdecodepercycle, \const¬maxstartpercycle, \tup{4, 4, 4, 1, 1}, \sequence{}}

Reviewer: ```suggestion
      &\gas¬sim¬state^{\text{init}} &=&\text{ }\tup{\imath, 0, \const¬maxdecodepercycle, \const¬maxstartpercycle, \tup{4, 4, 4, 1, 1}, \sq{}}
```

  }
\end{equation}

The gas cost model simulation function $\gas¬sim$ is defined as follows:

\begin{equation}
  \gas¬sim\colon \abracegroup{
    \tup{\blob, \bitstring, \gas¬sim¬ty} &\to \gas¬sim¬ty \\
    \tup{\mathbf{c}, \mathbf{k}, \gas¬sim¬state} &\mapsto \begin{cases}
      \gas¬sim(\mathbf{c}, \mathbf{k}, \gas¬sim'(\mathbf{c}, \mathbf{k}, \gas¬sim¬state)) &\when \gas¬sim¬state_{\imath} \neq \none \land \cost¬decodeslots(\mathbf{c}, \mathbf{k}, \gas¬sim¬state_{\imath}) \leq \gas¬sim¬state_{\remaining¬decodeslots} \land l < 32 \\
      \gas¬sim(\mathbf{c}, \mathbf{k}, \gas¬sim''(\gas¬sim¬state)) &\otherwhen \find¬readytostart(\gas¬sim¬state) \neq \none \land \gas¬sim¬state_{\remaining¬startpercycle} > 0 \\
      \gas¬sim¬state &\otherwhen \gas¬sim¬state_{\imath} = \none \land l = 0 \\
      \gas¬sim(\mathbf{c}, \mathbf{k}, \gas¬sim'''(\gas¬sim¬state)) &\otherwise \\
    \end{cases}\\

>     \where l &= |\set{\build{i}{ i \in \N, \in¬rob¬field{i}{\rob¬state} \neq \none }}|

Reviewer: Not sure it's really defined what `s[i]` means when `i >= |s|`. Here you can avoid this by writing eg:

    \len{\sq{\build{r}{ r \orderedin \in¬rob, r_{\rob¬state} \neq \none }}}

  }
\end{equation}

The state transition function $\gas¬sim'$ which decodes the instructions without triggering the virtual CPU pipeline simulation is defined as follows:

\begin{equation}
  \gas¬sim'\colon \abracegroup{
    \tup{\blob, \bitstring, \gas¬sim¬ty} &\to \gas¬sim¬ty \\
    \tup{\mathbf{c}, \mathbf{k}, \gas¬sim¬state} &\mapsto \begin{cases}
      \gas¬sim^{\text{mov}}(\mathbf{c}, \mathbf{k}, \gas¬sim¬state) &\when \text{opcode}(\mathbf{c}, \mathbf{k}, \gas¬sim¬state_{\imath}) = \token{move\_reg} \\
      \gas¬sim^{\text{dec}}(\mathbf{c}, \mathbf{k}, \gas¬sim¬state) &\otherwise
    \end{cases}
  }
\end{equation}

The \token{move\_reg} instruction is special-cased to be handled by the frontend of our virtual CPU, without being added to the reorder buffer:

\begin{equation}
  \gas¬sim^{\text{mov}}\colon \abracegrouptwo{
    \tup{\blob, \bitstring, \gas¬sim¬ty} &\to \gas¬sim¬ty \\
    \tup{\mathbf{c}, \mathbf{k}, \gas¬sim¬state} &\mapsto \gas¬sim¬state'
  }{

>     \where \gas¬sim¬state' &= \gas¬sim¬state\text{ except the following:}\\

Reviewer: Don't think you need "the following", the rest of the GP just uses "except:" and I think that reads more clearly.

      \gas¬sim¬state'_{\imath} &= \gas¬sim¬state_{\imath} + 1 + \Fskip(\gas¬sim¬state_{\imath}) \\
      \gas¬sim¬state'_{\remaining¬decodeslots} &= \remaining¬decodeslots - 1 \\
      \gas¬sim¬state'_{\rob} &= \in¬rob\text{ except the following:}\\
      &\begin{aligned}
        \rob¬entry¬field{\gas¬sim¬state'_{\rob}}{j}{\rob¬regs} &= \begin{cases}
          \in¬rob¬field{j}{\rob¬regs} \cup \inst¬dstregs(\mathbf{c}, \mathbf{k}, \gas¬sim¬state_{\imath}) &\when \in¬rob¬field{j}{\rob¬regs} \cap \inst¬srcregs(\mathbf{c}, \mathbf{k}, \gas¬sim¬state_{\imath}) \neq \emptyset \\
          \in¬rob¬field{j}{\rob¬regs} \setminus \inst¬dstregs(\mathbf{c}, \mathbf{k}, \gas¬sim¬state_{\imath}) &\otherwise
        \end{cases} \\
      \end{aligned} \\

>       j &\in \N

Reviewer: I think you want to write this like:

    forall j in N_|x_r| : x'_r[j]_r = ...

So quantify j properly and limit to just the valid indices.

  }
\end{equation}

Every other instruction is fully decoded and added to the reorder buffer as follows:

\begin{equation}
  \gas¬sim^{\text{dec}}\colon \abracegrouptwo{
    \tup{\blob, \bitstring, \gas¬sim¬ty} &\to \gas¬sim¬ty \\
    \tup{\mathbf{c}, \mathbf{k}, \gas¬sim¬state} &\mapsto \gas¬sim¬state'
  }{
    \where \gas¬sim¬state' &= \gas¬sim¬state\text{ except the following:}\\
    \gas¬sim¬state'_{\imath} &= \begin{cases}
      \none &\when \text{opcode}(\mathbf{c}, \mathbf{k}, \gas¬sim¬state_{\imath}) \in T \\
      \gas¬sim¬state_{\imath} + 1 + \Fskip(\gas¬sim¬state_{\imath}) &\otherwise
    \end{cases}\\
    \gas¬sim¬state'_{\remaining¬decodeslots} &= \remaining¬decodeslots - \cost¬decodeslots(\mathbf{c}, \mathbf{k}, \gas¬sim¬state_{\imath}) \\

>     \rob¬entry{\rob'}{j} &= \in¬rob¬entry{j}\text{ except: } \rob¬entry¬field{\rob'}{j}{\rob¬regs} = \in¬rob¬field{j}{\rob¬regs} \setminus \rob¬regs'\\

Reviewer: Should write this like:

    r' = x_r except: forall j in N_|x_r|: r'[j]_r = ...

    j &\in \N \\

>     \rob¬regs' &= \inst¬dstregs(\mathbf{c}, \mathbf{k}, \gas¬sim¬state_{\imath}) \\

Reviewer: Can probably remove the 's from these `rob` variables here.

    \rob¬cyclesleft' &= \cost¬cycles(\mathbf{c}, \mathbf{k}, \gas¬sim¬state_{\imath}) \\
    \rob¬execunits' &= \cost¬execunits(\mathbf{c}, \mathbf{k}, \gas¬sim¬state_{\imath}) \\

>     \rob¬deps' &= \set{\build{i}{ i \in N, \inst¬srcregs(\mathbf{c}, \mathbf{k}, \gas¬sim¬state_{\imath}) \cap \in¬rob¬field{i}{\rob¬regs} \neq \emptyset }} \\

Reviewer: ```suggestion
    \rob¬deps' &= \set{\build{i}{ i \in \N_{\len{\in¬rob}}, \inst¬srcregs(\mathbf{c}, \mathbf{k}, \gas¬sim¬state_{\imath}) \cap \in¬rob¬field{i}{\rob¬regs} \neq \emptyset }} \\
```

>     \gas¬sim¬state'_{\rob} &= \rob' \concat \tuple{\state¬decoding, \rob¬cyclesleft', \rob¬deps', \rob¬regs', \rob¬execunits'} \\

Reviewer: ```suggestion
    \gas¬sim¬state'_{\rob} &= \rob' \append \tuple{\state¬decoding, \rob¬cyclesleft', \rob¬deps', \rob¬regs', \rob¬execunits'} \\
```

  }
\end{equation}

The state transition function $\gas¬sim''$ which starts the execution of the next pending instruction is defined as follows:

\begin{equation}
  \gas¬sim''\colon \abracegrouptwo{
    \gas¬sim¬ty &\to \gas¬sim¬ty \\
    \gas¬sim¬state &\mapsto \gas¬sim¬state'\\
  }{
    \where \gas¬sim¬state' &= \gas¬sim¬state\text{ except the following:}\\
    \gas¬sim¬state'_{\remaining¬startpercycle} &= \gas¬sim¬state_{\remaining¬startpercycle} - 1\\
    n &= \find¬readytostart(\gas¬sim¬state)\\
    \gas¬sim¬state'_{\remaining¬execunits} &= \gas¬sim¬state_{\remaining¬execunits} - \in¬rob¬field{n}{\rob¬execunits}\\

>     \gas¬sim¬state'_{\rob} &= \in¬rob\text{ except: } \rob¬entry¬field{\rob'}{n}{\rob¬state} = \state¬executing

Reviewer: ```suggestion
    \gas¬sim¬state'_{\rob} &= \in¬rob\text{ except: } \rob¬entry¬field{\gas¬sim¬state'_{\rob}}{n}{\rob¬state} = \state¬executing
```

  }
\end{equation}

The function $\find¬readytostart$ which checks which instruction inside of the reorder buffer is ready to start executing (and whether such an instruction even exists) is defined as follows:

\begin{equation}
  \find¬readytostart\colon \abracegroup{

>     \gas¬sim¬ty &\to \set{\N, \none}\\

Reviewer: ```suggestion
    \gas¬sim¬ty &\to \N \cup \set{\none}\\
```

>     \gas¬sim¬state &\mapsto \min(j \in \N : \in¬rob¬field{j}{\rob¬state} = \state¬waiting \land \in¬rob¬field{j}{\rob¬execunits} \leq \gas¬sim¬state_{\remaining¬execunits} \land (\forall k \in \in¬rob¬field{j}{\rob¬deps} \Rightarrow \in¬rob¬field{k}{\rob¬cyclesleft} \leq 0))

Reviewer: I guess this is assuming `\min(\none) = \none`, not sure that is really defined anywhere though. I would probably just be explicit about this here for clarity, maybe something like:

    x -> \none if r = \none, min(r) otherwise
    where r = { j in \N_{\len{x_r}} | ... }

Separately, I don't think `\Rightarrow` should be used here. That is used to mean "implies" elsewhere, think you just want a colon here?

  }
\end{equation}

The state transition function $\gas¬sim'''$ which simulates the rest of the virtual CPU pipeline is defined as follows:

\begin{equation}
  \gas¬sim'''\colon \abracegrouptwo{
    \gas¬sim¬ty &\to \gas¬sim¬ty \\
    \gas¬sim¬state &\mapsto \gas¬sim¬state'\\
  }{
    \where \gas¬sim¬state' &= \gas¬sim¬state\text{ except the following:}\\
    \gas¬sim¬state'_{\remaining¬decodeslots} &= \const¬maxdecodepercycle, \gas¬sim¬state'_{\remaining¬startpercycle} = \const¬maxstartpercycle, \gas¬sim¬state'_{\ctr¬cycles} = \gas¬sim¬state_{\ctr¬cycles} + 1\\

>     \gas¬sim¬state'_{\remaining¬execunits} &= \gas¬sim¬state_{\remaining¬execunits} + \sum_{n \in \N \Rightarrow \in¬rob¬field{n}{\rob¬state} = \state¬executing \land \in¬rob¬field{n}{\rob¬cyclesleft} = 1} \in¬rob¬field{j}{\rob¬execunits} \\

Reviewer: ```suggestion
    \gas¬sim¬state'_{\remaining¬execunits} &= \gas¬sim¬state_{\remaining¬execunits} + \sum_{n \in \N_{\len{\in¬rob}}, \in¬rob¬field{n}{\rob¬state} = \state¬executing \land \in¬rob¬field{n}{\rob¬cyclesleft} = 1} \in¬rob¬field{n}{\rob¬execunits} \\
```

>     \rob¬entry{\gas¬sim¬state'_{\rob}}{j} &= \in¬rob¬entry{j}\text{ except the following:} \\

Reviewer: ```suggestion
    \gas¬sim¬state'_{\rob} &= \in¬rob\text{ except } \forall j \in \N_{\len{\in¬rob}}: \\
```

(And drop the `j \in \N` at the bottom)

    &\begin{aligned}
      \rob¬entry¬field{\gas¬sim¬state'_{\rob}}{j}{\rob¬state} &= \begin{cases}

>         \none &\when \forall k \in \N,\quad 0 \leq k \leq j \Rightarrow \in¬rob¬field{k}{\rob¬state} = \state¬executed \\

Reviewer: ```suggestion
        \none &\when \forall k \in \N_{j+1},\quad \in¬rob¬field{k}{\rob¬state} \in \set{\state¬executed, \none} \\
```

        \state¬waiting &\when \in¬rob¬field{j}{\rob¬state} = \state¬decoding \\

>         \state¬executed &\when \in¬rob¬field{j}{\rob¬state} = \state¬executing \land \in¬rob¬field{j}{\rob¬cyclesleft} = 0 \\

Reviewer: Is it intended that we stay in the EXE state for 1 extra cycle? Doesn't seem like this changes much other than how long it takes for buffer slots to free up, but it does seem a bit weird. In particular cycles will go negative but it's supposed to be `\in \N`.

Author: > Is it intended that we stay in the EXE state for 1 extra cycle?

Yes, this is intended as what matters are the transitions when the cycle counter ticks over.

Let's say we have an instruction with a cycle count of 1 cycle; triggering the state machine we get:

- Step 0: cycle 0 -> 0: instruction gets decoded, state is now DEC
- Step 1: cycle 0 -> 1: instruction is marked as waiting, state is transitioned DEC -> WAIT
- Step 2: cycle 1 -> 1: instruction is marked for execution, state is transitioned WAIT -> EXE
- Step 3: cycle 1 -> 2: cycle count is decremented and becomes 0, state unchanged EXE -> EXE
- Step 4: cycle 2 -> 3: cycle count is decremented and becomes -1, state is transitioned EXE -> FIN
- Step 5: cycle 3 -> 4: state is now FIN -> retired

Or if we only look at the cycle transitions start-to-end:

- Cycle 0 -> 1: state went DEC -> EXE
- Cycle 1 -> 2: state went EXE -> EXE
- Cycle 2 -> 3: state went EXE -> FIN
- Cycle 3 -> 4: state went FIN -> retired

So, it gets decoded, then executed, then marked as finished, then it's retired. What matters is for how many cycle *transitions* it *stays* as EXE (which is only one here, which I suppose intuitively makes sense because CPUs do "work" only when the clock ticks).

We can also put this in a timeline graph form, assigning each type of transition a letter:

```
0123
DeER
```

And here's what LLVM's microarchitectural simulator prints out if I ask it to simulate a 1-cycle instruction like `add`:

```
[0,0]     DeER .    . .   add	rax, rbx
```

...which is not a coincidence. (: 

> In particular cycles will go negative but it's supposed to be `\in \N`.

Ah yeah, good point. Practically this doesn't matter, so I suppose I can just clamp them?

Reviewer: > Practically this doesn't matter, so I suppose I can just clamp them?

Sure.

        \in¬rob¬field{j}{\rob¬state} &\otherwise \\
      \end{cases} \\
      \rob¬entry¬field{\gas¬sim¬state'_{\rob}}{j}{\rob¬cyclesleft} &= \begin{cases}
        \in¬rob¬field{j}{\rob¬cyclesleft} - 1 &\when \in¬rob¬field{j}{\rob¬state} = \state¬executing \\
        \in¬rob¬field{j}{\rob¬cyclesleft} &\otherwise \\
      \end{cases} \\

>       \rob¬entry¬field{\gas¬sim¬state'_{\rob}}{j}{\rob¬regs} &= \begin{cases}

Reviewer: This bit doesn't seem necessary, dependencies that have no cycles left have no effect?

Author: Hmm... ah yep, you're right. Good catch! Yeah, this is here because I originally wrote these equations based on my code where I *do* need to do this part, but after simplifying everything for the GP this is now not necessary.

    (\N, \N, \blob, \bitstring, \pvmreg) &\to \N\\
    (a, b, \mathbf{c}, \mathbf{k}, \imath) &\mapsto \begin{cases}
      a &\when \inst¬srcregs(\mathbf{c}, \mathbf{k}, \imath) \cap \inst¬dstregs(\mathbf{c}, \mathbf{k}, \imath) \neq \emptyset \\
      b &\otherwise
    \end{cases}
  }
\end{equation}

For non-immediate shift and rotate instructions only the first source register matters:

\begin{equation}
  \mathfrak{P}_{S}\colon \abracegrouptwo{
    (\N, \N, \blob, \bitstring, \pvmreg) &\to \N\\
    (a, b, \mathbf{c}, \mathbf{k}, \imath) &\mapsto \begin{cases}
      a &\when \reg_A = \reg_D \\
      b &\otherwise
    \end{cases}\\
  }{
    \\[0.2pt]
    &\where (\mathbf{c}, \mathbf{k}, \imath) \mapsto (\reg_A, \reg_D) \text{ according to \ref{sec:programdecoding} and \ref{sec:instructiontables}} \\
  }
\end{equation}

% NOTE: This is a *temporary* placeholder. We'll later replace this with a proper equation
%       once we introduce gas metering of the paging mechanism and memory limits,
%       so that's why I'm not inlining this into the table.
The cost of memory accesses is defined as follows:

\begin{equation}
  \mathfrak{m} \equiv 25
\end{equation}

> The cost of a branch depends on whether any of its targets (either the jump target or the implicit fallthrough) points to an instruction byte which is equal to the opcode for the \token{unlikely} or the \token{trap} instruction; formally:

Reviewer: ```suggestion
The cost of a branch depends on whether any of its targets (either the jump target or the implicit fallthrough) point to an instruction byte which is equal to the opcode for the \token{unlikely} or the \token{trap} instruction; formally:
```

      #2
    \end{aligned}
  \end{array}
  \right.
}

The gas cost model for the \textsc{pvm} is a simplified model of a modern CPU microarchitecture, heavily inspired by what's used by production-grade compilers to predict how much time a given piece of code will take. For each basic block in the program the model simulates its execution flow and computes the required number of virtual CPU cycles that would be needed to execute it.

A gas simulation state, of set $\gas¬sim¬ty$, contains the current instruction counter $\imath$, a cycle counter $\ctr¬cycles$, the number of decode slots $\remaining¬decodeslots$ remaining in the current cycle, the maximum number of instructions $\remaining¬startpercycle$ which are still allowed to start execution in the current cycle, the remaining available execution units $\remaining¬execunits$, and the reorder buffer $\rob$. Formally:

\begin{equation}
  \gas¬sim¬ty \equiv \tuple{
    \isa{\imath}{\pvmreg \cup \set{\none}},
    \isa{\ctr¬cycles}{\N},
    \isa{\remaining¬decodeslots}{\N},
    \isa{\remaining¬startpercycle}{\N},
    \isa{\remaining¬execunits}{\execunits¬ty},
    \isa{\rob}{\sequence{\rob¬entry¬ty}}
  }\\
\end{equation}

> A reorder buffer entry state, of set $\rob¬entry¬ty$, contains its current state $\rob¬state$, the number of cycles left $\rob¬cyclesleft$, a set of reorder buffer indices considered its dependencies $\rob¬deps$, a set of clobbered registers $\rob¬regs$, and the execution units used $\rob¬execunits$. Formally:

Reviewer: ```suggestion
A reorder buffer entry, of set $\rob¬entry¬ty$, contains its current state $\rob¬state$, the number of cycles left $\rob¬cyclesleft$, a set of reorder buffer indices considered its dependencies $\rob¬deps$, a set of clobbered registers $\rob¬regs$, and the execution units used $\rob¬execunits$. Formally:
```

\begin{equation}
  \rob¬entry¬ty \equiv \tuple{
    \isa{\rob¬state}{\rob¬state¬ty},
    \isa{\rob¬cyclesleft}{\N},
    \rob¬deps \subseteq \N,
    \rob¬regs \subseteq \Nmax{13},
    \isa{\rob¬execunits}{\execunits¬ty}
  }\\
\end{equation}

> A set $\execunits¬ty$ maps each execution unit kind (${\ALU, \LOAD, \STORE, \MUL, \DIV}$) into a number:

Reviewer: ```suggestion
A tuple of the set $\execunits¬ty$ maps each execution unit kind (${\ALU, \LOAD, \STORE, \MUL, \DIV}$) into a number:
```

\begin{equation}
  \execunits¬ty \equiv \tuple{
    \isa{\ALU}{\N},
    \isa{\LOAD}{\N},
    \isa{\STORE}{\N},
    \isa{\MUL}{\N},
    \isa{\DIV}{\N}
  }
\end{equation}

> For convenience we define addition and subtraction of two sets $\execunits¬ty$ to be equivalent to the memberwise operation of the same kind; formally:

Reviewer: ```suggestion
For convenience we define addition and subtraction of two tuples of the set $\execunits¬ty$ to be equivalent to the memberwise operation of the same kind; formally:
```

\begin{equation}
\begin{aligned}
  \forall a,b \in \execunits¬ty;\forall \oplus \in \set{+, -}&: a \oplus b \equiv c\\
  \where c \in \execunits¬ty, \forall k \in \set{\ALU, \LOAD, \STORE, \MUL, \DIV}&:c_k \equiv a_k \oplus b_k\\
\end{aligned}
\end{equation}

> Similarly, a comparison between two sets $\execunits¬ty$ is true if and only if the chosen comparison holds for all corresponding members:

Reviewer: ```suggestion
Similarly, a comparison between two tuples of the set $\execunits¬ty$ is true if and only if the chosen comparison holds for all corresponding members:
```

We also define the function $\cost¬cycles(\mathbf{c}, \mathbf{k}, \imath)$ which returns the number of cycles the instruction at $\imath$ needs to finish execution, $\cost¬decodeslots(\mathbf{c}, \mathbf{k}, \imath)$ which returns the number of decoding slots necessary to decode it, and $\cost¬execunits(\mathbf{c}, \mathbf{k}, \imath)$ which returns the number of virtual CPU execution units required to start its execution. These simply return the values as specified in \ref{sec:gascosttables}.

The gas cost of a given basic block starting at instruction opcode index $\imath \in \basicblocks$ and given the instruction data $\mathbf{c}$ and the opcode bitmask $\mathbf{k}$ is defined by the number of virtual CPU cycles as determined by the gas cost model transition function, up until every instruction of the basic block it has ingested has been retired and the simulation has converged. Formally:

\begin{equation}\label{eq:gascostforblock}
  \gascostforblock\colon \abracegrouptwo{
    \tup{\blob, \bitstring, \pvmreg} &\to \N\\

>     \tup{\mathbf{c}, \mathbf{k}, \imath} &\mapsto \max(\gas¬sim¬state^{\text{final}}_{\ctr¬cycles} - 3, 1)\\

Reviewer 2: The `max(cycles - 3, 1)` in the final gas cost gives a 3-cycle "free" startup, correct? I'm just curious why?

Author: In a nutshell: programs are very often full of tiny basic blocks, and real CPUs can "see" across those; our gas cost model can't. Considering the rest of the model pretty much assumes the worst case, this optimistically tries to claw back a little bit of gas.

That said, as with the rest of the numbers in the model, this value is subject to change and might be either removed or reduced to zero if it ends up being insecure after more thorough testing.

  }
\end{equation}

The state transition function $\gas¬sim'$ which decodes the instructions without triggering the virtual CPU pipeline simulation is defined as follows:

\begin{equation}
  \gas¬sim'\colon \abracegroup{
    \tup{\blob, \bitstring, \gas¬sim¬ty} &\to \gas¬sim¬ty \\
    \tup{\mathbf{c}, \mathbf{k}, \gas¬sim¬state} &\mapsto \begin{cases}
      \gas¬sim^{\text{mov}}(\mathbf{c}, \mathbf{k}, \gas¬sim¬state) &\when \imath < \len{\mathbf{c}} \land \mathbf{c}_{\imath} = \token{move\_reg} \\
      \gas¬sim^{\text{dec}}(\mathbf{c}, \mathbf{k}, \gas¬sim¬state) &\otherwise
    \end{cases}
  }
\end{equation}

The \token{move\_reg} instruction is special-cased to be handled by the frontend of our virtual CPU, without being added to the reorder buffer:

\begin{equation}
  \gas¬sim^{\text{mov}}\colon \abracegrouptwo{
    \tup{\blob, \bitstring, \gas¬sim¬ty} &\to \gas¬sim¬ty \\
    \tup{\mathbf{c}, \mathbf{k}, \gas¬sim¬state} &\mapsto \gas¬sim¬state'
  }{
    \where \gas¬sim¬state' &= \gas¬sim¬state\text{ except:}\\
      \gas¬sim¬state'_{\imath} &= \gas¬sim¬state_{\imath} + 1 + \Fskip(\gas¬sim¬state_{\imath}) \\

>       \gas¬sim¬state'_{\remaining¬decodeslots} &= \remaining¬decodeslots - 1 \\

Reviewer 2: ```suggestion
      \gas¬sim¬state'_{\remaining¬decodeslots} &= \gas¬sim¬state_{\remaining¬decodeslots} - 1 \\
```

      &\begin{aligned}
        \forall j &\in \in¬rob¬len : \rob¬entry¬field{\gas¬sim¬state'_{\rob}}{j}{\rob¬regs} &= \begin{cases}
          \in¬rob¬field{j}{\rob¬regs} \cup \inst¬dstregs(\mathbf{c}, \mathbf{k}, \gas¬sim¬state_{\imath}) &\when \in¬rob¬field{j}{\rob¬regs} \cap \inst¬srcregs(\mathbf{c}, \mathbf{k}, \gas¬sim¬state_{\imath}) \neq \emptyset \\
          \in¬rob¬field{j}{\rob¬regs} \setminus \inst¬dstregs(\mathbf{c}, \mathbf{k}, \gas¬sim¬state_{\imath}) &\otherwise
        \end{cases} \\
      \end{aligned}
  }
\end{equation}

Every other instruction is fully decoded and added to the reorder buffer as follows:

\begin{equation}
  \gas¬sim^{\text{dec}}\colon \abracegrouptwo{
    \tup{\blob, \bitstring, \gas¬sim¬ty} &\to \gas¬sim¬ty \\
    \tup{\mathbf{c}, \mathbf{k}, \gas¬sim¬state} &\mapsto \gas¬sim¬state'
  }{
    \where \gas¬sim¬state' &= \gas¬sim¬state\text{ except:}\\
    \gas¬sim¬state'_{\imath} &= \begin{cases}
      \none &\when \instructions_{\imath} \in T \\
      \gas¬sim¬state_{\imath} + 1 + \Fskip(\gas¬sim¬state_{\imath}) &\otherwise
    \end{cases}\\
    \mathbf{c} &\mapsto \instructions \text{ according to \ref{eq:instructions}}\\

>     \gas¬sim¬state'_{\remaining¬decodeslots} &= \remaining¬decodeslots - \cost¬decodeslots(\mathbf{c}, \mathbf{k}, \gas¬sim¬state_{\imath}) \\

Reviewer 2: ```suggestion
    \gas¬sim¬state'_{\remaining¬decodeslots} &= \gas¬sim¬state_{\remaining¬decodeslots} - \cost¬decodeslots(\mathbf{c}, \mathbf{k}, \gas¬sim¬state_{\imath}) \\
```
